### DISCUSSION TITLE
"Supposed bad word" makes copilot stop working

### DISCUSSION THREAD

**Discussion by JuanBerta** [community/community#7079] [Copilot Conversations]
I don't know where to ask for this, if someone can move it to the correct place I would be grateful.

I have a problem with using copilot, using the word "sex" even if it it means "genre" in the code it don't recognizes it as genre and copilot stops autocompleting code.

Code:

import pandas as pd
import numpy as np

train_df = pd.read_csv(
    'Titanic dataset/train.csv')
test_df = pd.read_csv(
    'Titanic dataset/test.csv')

print(train_df.head())

# prepare data
# drop unnecessary columns, these columns won't be useful in analysis and prediction
train_df = train_df.drop(['PassengerId', 'Name', 'Ticket'], axis=1)
test_df = test_df.drop(['Name', 'Ticket'], axis=1)

# fill nan values of age
train_df['Age'].fillna(train_df['Age'].mean(), inplace=True)

# fill nan values of cabin
train_df['Cabin'].fillna('U', inplace=True)

# fill nan values of embarked
train_df['Embarked'].fillna('S', inplace=True)

# fill nan values of fare
train_df['Fare'].fillna(
    train_df['Fare'].mean(), inplace=True)

# fill nan values of age in test data
test_df['Age'].fillna(test_df['Age'].mean(), inplace=True)

# fill nan values of cabin in test data
test_df['Cabin'].fillna('U', inplace=True)

# fill nan values of embarked in test data
test_df['Embarked'].fillna('S', inplace=True)

# fill nan values of fare in test data
test_df['Fare'].fillna(test_df['Fare'].mean(), inplace=True)

# map embarked to int
train_df['Embarked'] = train_df['Embarked'].map(
    {'S': 0, 'C': 1, 'Q': 2}).astype(int)

# map
test_df['Embarked'] = test_df['Embarked'].map(
    {'S': 0, 'C': 1, 'Q': 2}).astype(int)

# map train_df cabin to int
train_df['Cabin'] = train_df['Cabin'].map(
    {'A': 0, 'B': 1, 'C': 2, 'D': 3, 'E': 4, 'F': 5, 'G': 6, 'T': 7, 'U': 8}).astype(int)

# map
test_df['Cabin'] = test_df['Cabin'].map(
    {'A': 0, 'B': 1, 'C': 2, 'D': 3, 'E': 4, 'F': 5, 'G': 6, 'T': 7, 'U': 8}).astype(int)

# train_df['Sex'] to int (here stops working)

**Comment by MulleDK19** [comment]
Same with gender...

**Comment by Benjaminbburn** [comment]
I can see why they might have this but it seems to be a bit heavy handed. 

I see two useful options: 
1. Allow the user to take a deliberate action to disable it all together
2. If word is deemed to be "inappropriate", simply redact it from the suggestion. 

Example)
(in python)
(Trying to trigger a suggestion via comment)
(using the heart disease dataset which uses "sex" to refer to Male/Female)

\# create bar plot of sex and target  -> **this results in no suggestion**
Ideally, it would result in: **df.plot(kind='bar', x='sex', y='thal')**

I would at least like to see: **df.plot(kind='bar', x=, y='thal')  #Output modified due to restricted word**

**Comment by nephi-dev** [comment]
Same with "comp_man" that acctualy means "Company Manager". I don't even know why is copilot stoping in the word "man", it auto-complete only to "comp_". A little bit strange.

**Reply by MulleDK19** [reply]
Same with woman..

This cannot be completed.

```csharp
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;

internal static class Program
{
    private static void Main()
    {
        Console.Write("Are you a m4n or woman (M/W)? ");
        ConsoleKeyInfo|
    }
}
```

... doesn't help that there's NO feedback when this happens. You just get nothing..

If your code contains any words like sex, gender, man, woman, etc. Copilot becomes useless..

**Comment by nephi-dev** [comment]
An option to disable this type of filter will be great.

**Reply by MulleDK19** [reply]
Not gonna happen, sadly..

**Comment by cmp-nct** [comment]
It's similar to what happened with Dall-E2 after Open Diffusion went public.
Dall-E does have a superior model, they likely didn't crop their training data's heads, legs and fingers off.
However "no celebrities, no faces, no offending content".
What happened ? it died within a week. Well it's still there but 95% of the people are gone forever.
Also SD was optimized within a month to a point that made it 2-3 times faster and from 12GB of minimal GPU memory it went down to 4GB without loss of quality.
Dall-E freaked out and changed many policies, people can use faces, they can upload photos again .. that won't save the business from returning to what it could have become. SD is progressing too fast with thousands of developers working on hundreds of forks within 2 months of starting out.

The same will likely happen to copilot if it tries to change how people think or what they code.
No serious developer will accept that he can't use "sex" or "gender" in his projects.

**Comment by mtanksley** [comment]
Even the AI is becoming "woke" in today's society. Everything has become censored and neutered and AI has been no exception. I'd imagine in the future Copilot will be auto-reporting developers to the authorities for moral and ethical offenses in their code.

**Reply by nephi-dev** [reply]
just give some time to stability and a Open Souce LLM will come out, it's just a matter time now

**Comment by kellynauert** [comment]
Also filters out race and ethnicity. Very frustrating.

**Comment by mniederl** [comment]
This is even more annoying when it concerns other languages with overlapping words.
For example, after "Wie dick ist das Blatt?" in German ("How thick is the sheet?") nothing more is suggested although the meaning of the second word in English is completely irrelevant for German.

**Comment by cmp-nct** [comment]
Though to be realistic, the time of copilot is already ending. The opensource models are not far behind and 1-2 generations further they will be ahead. Which means all your code belongs to you again and you do not have to censor comments to the language someone else wants to force you and your team into.
I expect that even master-slave terminology will be blocked one day for something like "elected principal/voluntary worker"

**Reply by og-buzz** [reply]
I am writing a master-slave modbus program and yeah this is the case currently.

**Comment by iam-amanxz** [comment]
gender is a bad word in 2023 sadly

**Reply by Weinwurm** [reply]
Now imagine you are working on medical data! We are using now `bender = 'g' + 'e' + 'n' + 'd' + 'e' + 'r'`, but will move away from Copilot Big Nanny asap.

**Comment by gusbemacbe** [comment]
People,

I suggest you try CodeGeeX. It does not censor the terms, as I tested.

**Reply by iam-amanxz** [reply]
Have you tried it, is it as good as Copilot?

**Reply by gusbemacbe** [reply]
I have used for 2 months. CodeGeeX still doesn't beat/overcome/surpass Github Copilot because occasionally, it misunderstood what you tried to write and wrote the wrong codes, but produced some good results in JavaScript and in Python, when I wrote the conditionals and the regular expressions. CodeGeeX has only a big advantage: it does not censor any word. 

You can test and try on https://codegeex.cn/playground before you install the extension and see if it is good for you.

**Reply by iam-amanxz** [reply]
sure mate, thanks

**Comment by krisgrm** [comment]
same issue here :( 

what I can say is that https://www.cursor.sh/ doesn't have the same issue, so seems like cursor should be used in favour of copilot

**Reply by Lamby777** [reply]
Switching your whole editor just so the AI model doesn't try to police your vocabulary is a pretty big jump to make...

**Comment by whalderman** [comment]
Similar filters are not applied to other languages/locales. Why is that?  
Australian curse words appear to be fine. Is there a reason they (don't) get special treatment?  
Japanese curse words are fine too.

What was the intended goal here?  
Only a specific subset of English speakers appear to be affected by this (whether or not that's based on VSCode's locale only changes the context here).

**Comment by Lamby777** [comment]
I named one of my modules in a joke project `i_fucking_love_enums.rs` because I was doing weird 100% insane cursed stuff you should never do with enums and now it just refuses to suggest anything near that part of the code because of the name. It's pretty annoying, but I absolutely refuse to rename that module and let Copilot police my language like that. The moment I do is the moment the people who put this censorship in place get what they want.

**Comment by hebabz** [comment]
Just now I was doing a bunch of filters for a university project and when I put 2 and 2 together I guessed why no auto-complete for the gender one and honestly that made me chuckle. Had to check on Google if what I guessed was right and I arrived here. Thanks for the confirmation!

**Comment by cmp-nct** [comment]
It's just a temporary problem.
If you look at opensource solutions, there are a few now that deliver quality significantly better than copilot did just half a year ago.
The open source projects are catching up and it's just a matter of maybe a year or two until they overtake the best OpenAI has to offer.
At the same time openai AI degrades, I paid 100-200 bucks a month just for GPT-4 answers alongside copilot. Since a week that stopped because GPT4 (all variants) degarded significantly.
Maybe "safer" now, in any case it's more like turbo-3 by now.

**Reply by MulleDK19** [reply]
Unfortunately, the Open Source variants don't wok on VS either, and the one that does, doesn't either, and there's no support to get from the Open Source community.

**Reply by gusbemacbe** [reply]
@cmp-nct and @MulleDK19, have you already tried CodeGeeX? It doesn't censor the words Copilot censors.

**Reply by MulleDK19** [reply]
Yes. I didn't find it very good; and it also doesn't support Visual Studio. Codeium is supposed to work with Visual Studio, but it's broken and there's no support, so I haven't been able to test that.

**Comment by std4453** [comment]
With all these advanced AI that could 'understand' human beings, they chose the easiest and absolutely worst solution, which is word-based censorship. How hilarious.

**Comment by iam-amanxz** [comment]
I understand people have different opinions when it comes to this and I respect that. But we should be able to locally manage what we should and should not see. Like a config. I think it's fair.

**Comment by knkillname** [comment]
I am a professor at a local Mexican university. I am trying to teach Digital Image Processing with Python, and whenever I write code referring to the black color in spanish (*negro*) I need to change it to **#000000**; otherwise copilot stops working altogether.

**Comment by AlanElMago** [comment]
I am a university student from Mexico. I'm learning image processing with Python and it seems to stop working when I type in the word _negro_ (the color black in spanish). Funnily, it works if I just type _negr_.

**Reply by CetoBasilius** [reply]
I'm having the exact same problem with the spanish word "negro". I'm building a challenge platform for league of legends and it references stealing a black elixir, which in spanish is "Robar elixir negro" so copilot decided to stop there.

**Comment by DrFeelgood123154** [comment]
Didn't have this before but today copilot suddenly stopped working for me after about a year of use, we have a massive project and remembering every thing is impossible so I've been relying on copilot to do that for me. Today I wrote something like "this shit code needs to be moved after x" in a comment and it stopped working. Took me 20 minutes to figure out why, then I tested and noticed it doesn't work with certain words.

So now I have to find an alternative that isn't infected by wokenism and will function properly as it is expected to function. This issue could also be considered fake advertising and end up with a court case.

**Reply by whalderman** [reply]
I've stopped using copilot at this point.

Funnily enough, programming is enjoyable again.

**Reply by DrFeelgood123154** [reply]
I would stop it too but sadly it saves me too much time and effort to be worth stopping it. It obviously can't do much - it's basically an advanced autocomplete but that is enough to give me like 15%-20% efficiency boost at work. I remember seeing "tabnine" as a viable alternative to copilot so will have to see how it does

**Comment by pingumen96** [comment]
When I use `tit` as abbreviation of `titolo` (italian for _title_, in my domain I need to give abbreviations of max 4 characters to each field), Copilot decides that's enough. Great job lol

**Comment by d4rky-pl** [comment]
Apparently `nationality` is also on the list of evil words you're not supposed to speak out loud.
What in the absolute f...

**Reply by Thiht** [reply]
I thought I was going crazy, but Copilot simply refuses to write the word `nationality`

Great for writing business code where the legal rules depend on the nationality of a user

**Comment by SKGleba** [comment]
can't use "slave" either, pain

**Comment by bojanpotocnik** [comment]
Working with PlatformIO/Arduino project in CLion - it uses I2C bus, which has clearly defined `master` and `slave` device terminology, rendering Copilot useless.
Although these might be "strong" filter words, reading other examples in this thread is just absurd. I am grateful that I can still use the word "color" or even specific color names 🙏

**Comment by cmp-nct** [comment]
In latest a year from now there should be very reliable open plugins available that replace copilot. Right now the integration of stuff like Continue.dev is not refined enough for professional use. But they are in a losing battle, no one will use censored AIs if the free or cheap ones work as well and are not censored or politically sided. 
It's as if Microsoft tries to change what people are working on

**Comment by doogie99** [comment]
Yep - I'm working on Veterinary EMR integrations and when I need to use "sex" CoPilot code generation just seizes. Ditto on asking it questions about master/slave relationships, which are very common in tech.

**Comment by icarosadero** [comment]
I have been noticing that Co-pilot stops giving suggestions when my code contains the word 'fuck' on it. The thing is that I am writing Brainfuck code samples, and it would be very handy to see its suggestions. Did they implement an option to disable whatever puristic filter they have implemented? I am paying for the damn thing, and should have the right to get suggestions regardless of whatever my variable names are.

**Reply by whalderman** [reply]
No way to disable it from what I recall.

**Comment by gusbemacbe** [comment]
@doogie99 and @icarosadero, did you try Cody, Fitten Code, Tabby, and Tabnine?

If you are totally frustrated with these alternatives, that I indicated... I have a suggestion. Of course, it is horrible and gives a lot of job. But it is worthy to try, and here are the instructions:

1. Construct a language Copilot can not speak and understand, for example, High Valyrian, Klingon, Kryptonian, Mandalorian, Na'vi, Quenya, or Vulcan or **even your own alien personal language** that Copilot never heard.
2. Create a text file containing a list of the native words on other text editor without Copilot for searching.
3. Create another text file containing a list of the alien words of your alien language on other text editor without Copilot for replacing, and to reverse. 
4. Write a script file in your favourite backend programming to create a function to get a specific file, and to search the words from the first text file and to replace them with the words from the second text file.

When you begin the project, use the alien words. When you finish the project, you need to close the files and run the terminal to run the script file.

When you need to change or fix the project, run the terminal to run the script file to reverse the words. After finishing the changes or the corrections, you run again the script file to replace again. Ready. 

Then Copilot can not detect nor mistrust the alien words that do not exist in all our real languages.

**Comment by Mancitiss** [comment]
I'm working with a list of video file names in PHP, everything went smoothly until I reach the file "Muder.on.the.Orient.Express.2017.mkv" then it stops working. Not only for that item but for all the next items in the list. It suggest NOTHING. I figure "Murder" was the word, changed it to "Love" then it starts working again, and it continue to suggest for all the next items in the list too. But the moment I changed it back to "Murder", even if I have long since passed that item, it suggests NOTHING.

**Comment by wolfmcnally** [comment]
I had to implement a hack in my language translation files. Can you guess which word Copilot didn't like?

```json
  "black": {
    "description": "The color black, used as fill color for elements.",
    "en_US": "Black",
    "es_ES": "!orgeN",
    "de_DE": "Schwarz",
    "fr_FR": "Noir",
    "ru_RU": "Чёрный",
    "hi_IN": "काला",
    "ko_KR": "검정",
    "th_TH": "สีดำ",
    "ja_JP": "黒",
    "zn_CN": "黑色",
    "he_IL": "שחור",
    "ar_SA": "أسود"
  },
```
